### PR TITLE
Simplify root directory

### LINKED ISSUE #145: Updating various dev-ops things to the new format in Nengo
Should we make the changes that `nengo/nengo` made? Specifically:

- Putting the requirements inside `setup.py` instead of the text files
- Refactoring the `.travis.yml` file into separate shell scripts
- Running `codespell -q 3` to avoid typos

Somewhat related to #143 and #96

### PR DESCRIPTION
**Pull request by Seanny123**
**Motivation and context:**
Resolves #145.

**How long should this take to review?**
- Average (neither quick nor lengthy)

**Types of changes:**
- Non-code change (touches things like tests, documentation, build scripts)

**Checklist:**

- [x] I have read the **CONTRIBUTING.rst** document.
- [ ] I have updated the documentation accordingly.
- [ ] I have included a changelog entry.
- [x] I have run the test suite locally and all tests passed.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -25,48 +25,48 @@ def read(*filenames, **kwargs):
 
 root = os.path.dirname(os.path.realpath(__file__))
 version_module = imp.load_source(
-    'version', os.path.join(root, 'nengo_spa', 'version.py'))
-testing = 'test' in sys.argv or 'pytest' in sys.argv
+    "version", os.path.join(root, "nengo_spa", "version.py"))
```
Comment (Seanny123) [inline setup.py]: Is this supposed to be changed from `nengo_spa` to `nengo-spa`?

Comment (tbekolay) [reply setup.py]: No, this refers to the `nengo_spa` directory relative to the root ([this one](https://github.com/nengo/nengo-spa/tree/master/nengo_spa))

**Comment by Seanny123** [comment]
It could, but I want to everything excluding the docs working first.

**Comment by jgosmann** [comment]
Probably makes most sense to include it in this PR because it will touch the same files.

**Comment by jgosmann** [comment]
Did you set the execute permissions on the shell scripts?

**Comment by tbekolay** [comment]
> Did you set the execute permissions on the shell scripts?

Yeah sorry, it's that. `chmod a+x .ci/*`

**Comment by Seanny123** [comment]
Whelp, turns out my mental model of file permissions being controlled by the OS was incorrect...

**Comment by Seanny123** [comment]
Tests are all passing now. Still need to:
- [ ] Upload documentation automatically
- [ ] Run codespell during static checks

File: .ci/docs.sh [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    echo
+    echo "  install  Install dependencies for building docs"
+    echo "  run      Build documentation"
+    echo "  upload   Upload documentation to gh-pages branch"
+    exit 1
+}
+
+if [[ "$COMMAND" == "install" ]]; then
+    conda install matplotlib scipy
+    pip install jupyter_client sphinx nbsphinx sphinx_rtd_theme ghp-import
```
Comment (jgosmann) [inline .ci/docs.sh]: Why are we installing some dependencies with conda and others with pip?
Also, do we have to hide the dependencies in this file? I'd prefer to properly list those in `setup.py` (as optional dependency where appropriate). If that does not work, a requirements file seems better to me (because than we are listing the dependency in two places anyways assuming setup.py lists them too and a requirements file is seems more discoverable).

File: .ci/docs.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    git push -fq "https://$GH_TOKEN@github.com/nengo/nengo-spa.git" gh-pages
+else
+    if [[ -z "$COMMAND" ]]; then
+        echo "Command required"
+    else
+        echo "Command $COMMAND not recognized"
+    fi
+    echo
+    usage
+fi
```
Comment (jgosmann) [inline .ci/docs.sh]: Add newline to end of file for consistency?

File: .travis.yml [hunk 4]
```diff
@@ -26,55 +25,41 @@ matrix:
...
     else
-      flake8 -v nengo_spa && pylint nengo_spa;
+      .ci/test.sh run;
     fi
 
 after_success:
   - if [[ "$COVERAGE" == "true" ]]; then
-      eval "bash <(curl -s https://codecov.io/bash)";
+      .ci/coverage.sh upload;
+    elif [[ -n "$TRAVIS_TAG" && "$DOCS" == "true" ]]; then
```
Comment (jgosmann) [inline .travis.yml]: Swap the conditions to be consistent with above?

File: MANIFEST.in [hunk 5]
```diff
@@ -15,11 +15,9 @@ prune docs/_build
...
 prune examples
 
-prune .*
-prune .git
-prune .tox
-
-prune nengo_spa/examples/.ipynb_checkpoints
+# Patterns to exclude from any directory
+global-exclude *.git*
+global-exclude *-checkpoint.ipynb
```
Comment (jgosmann) [inline MANIFEST.in]: Did you test this change?
Why did you remove `prune .*` (Though maybe it wasn't working because `.git` and `.tox` are explicitly listed?)
We probably should keep `prune .tox`.

File: setup.cfg [hunk 6]
```diff
@@ -12,18 +12,35 @@ all_files = 1
...
+markers =
+    example: Mark a test as an example.
+    noassertions: Mark a test without assertions. It will only be run if plots or analytics data are produced.
+    slow: Mark a test as slow to skip it per default.
+filterwarnings =
+    ignore::ImportWarning
+    ignore:(Buffer|Memory):DeprecationWarning
+    ignore:Skipping some optimization steps
+    ignore:SciPy is not installed
+ignore:numpy.(dtype|ufunc) size changed
```
Comment (jgosmann) [inline setup.cfg]: missing indentation?

File: setup.py [hunk 7]
```diff
@@ -25,48 +25,60 @@ def read(*filenames, **kwargs):
...
+        "nengo>=2.7",
+        "numpy>=1.8",
     ],
-    extras_require={
-        'prob_cleanup': ['scipy'],
+    extras_require = {
+        "all": docs_require + optional_requires + tests_require,
+        "docs": docs_require,
+        "optional": optional_requires,
+        "tests": tests_require
```
Comment (jgosmann) [inline setup.py]: Why are we no longer using the `tests_require` keyword argument?

Comment (tbekolay) [reply setup.py]: You should probably list test requirements in both an `extras_require` and `tests_require`, though note that the `tests_require` only comes into play if you do `python setup.py test` (and I think most of us don't do that).

File: .ci/docs.sh (hunk 2 above)
Comment (Seanny123) [reply .ci/docs.sh]: I literally have no idea. @tbekolay would you care to enlighten me and @jgosmann?

Comment (tbekolay) [reply .ci/docs.sh]: We use Conda with TravisCI to speed up builds significantly -- at least, that was the main intention before wheels came along. However, even now I believe the wheels situation on Linux is not great, so you're likely to be building NumPy from source if you `pip install numpy`. That takes a long time, and is not likely to give you a very efficient install of NumPy. Conda will give you a binary that uses MKL, so it's fast to install and fast to run.

NumPy is the main reason to use Conda, and then if you're already using Conda, you might as well use it for other packages that need to build C extensions (e.g. scipy) and packages with tricky installations (e.g. Matplotlib, or at least it used to be tricky, I think it's fine now).

You could replace the `pip install` line with a `pip install nengo[docs]` or something, but there are some things (like `ghp-import`) that are specific to TravisCI and are not actually required to build the docs. It's up to the Nengo SPA maintainer how exactly you want to organize these things, but for me, I see setting up TravisCI as a slightly different task than setting up a user's environment. For TravisCI, I want the minimal set of packages installed as quickly as possible to run the build job, and ideally I want the most recent version of those packages so that we can find any issues arising from updates in dependent packages. For a user's environment, I wouldn't want to prescribe that much and would rather keep things generic (e.g. if you want to build the docs, do `pip install nengo[docs]`); if they already have some older version of Sphinx installed, then that's most likely fine, and if not I can recommend upgrading.

Comment (Seanny123) [reply .ci/docs.sh]: It turns out you can now force `pip` to install binaries with the option `pip install --only-binary=:all:` and it seems to work well (on my machine). However, since `conda` was already doing the virtual environment management, removing this dependency would mean having to figure out a different way to do this, which might not be worth it...

Comment (tbekolay) [reply .ci/docs.sh]: If the tests are no slower using only pip, then it definitely makes sense to switch. I would try it and see. The virtual environment stuff doesn't matter on TravisCI because it is itself an isolated environment (a docker container); we don't have to use virtual environments (we only did in conda because everything is already in a virtual environment, essentially, so might as well).

**Comment by jgosmann** [comment]
@Seanny123 I think your last commit uses Travis CI's default Python instead of the requested Python version. [We don't necessarily need conda to manage Python version](https://docs.travis-ci.com/user/languages/python/), but specifying [the build matrix](https://docs.travis-ci.com/user/customizing-the-build#Build-Matrix) might be a little bit more complicated because it is quite sparse and will require a number of exclusions.

It might be the easiest to keep conda (above I wasn't intending to say that we should only use conda or pip, I just wanted to know why we're using both). Or at least to the Python install and virtualenv setup with conda and then install other dependencies with `pip`. Then I would try to install the base dependencies with `pip install nengo-spa[option]`. Dependencies that are specific for Travis CI (e.g. `ghp-import`) I would still list explicitely there. Either install them with `conda` or `pip`.

File: MANIFEST.in (hunk 5 above)
Comment (Seanny123) [reply MANIFEST.in]: My intention was to copy [core nengo's `MANIFEST.in`](https://github.com/nengo/nengo/blob/master/MANIFEST.in), but I might have done this incorrectly. Would you mind explaining to me the differences between this project's `MANIFEST.in` and core nengo's?

Comment (jgosmann) [reply MANIFEST.in]: The differences are probably not very meaningful. There are multiple ways to achieve the same result. I suppose I was using `prune` for `.git` and `.tox` because those directories should only appear in the root, but `global-exclude` might work too (I'm just not entirely sure whether it excludes directories or only files, that's why I was asking whether this change was tested).

There is a `global-exclude *.tox*` in the Nengo `MANIFEST.in` and even though we are removing the tox configuration some people might still use it or have an old `.tox` directory lieing around that should not be committed.

Comment (Seanny123) [reply MANIFEST.in]: Tested it and added the exclusion of `tox`. Looked good from visual inspection.

**Comment by jgosmann** [comment]
I added two small commits. Looks good to you?

**Comment by Seanny123** [comment]
They look fabulous! Merge at will.
